### PR TITLE
Adding real time validation to signup

### LINKED ISSUE #1433: Add Realtime Validation to Account Registration Page Realtime
The proposal for this issue is (as a progressive enhancement) to use js to show users real-time as-they-type feedback as to whether their username and/or email are available, and if their password meets the criteria.

## Sub-Tasks
The following issues should be created and resolved before this task may be closed.  
- [ ] #2053 - update backend API endpoints
- [ ] #2055 - Add real-time field validation for email, username, password to Account Registration Form front-end

## Reasoning
We recently had a regression #1431 to our account creation page which showed up as a server error when a user tried to register a new account but their username or email was already taken.

We spotted the error because our daily registered users dropped from ~2300 to ~1700 (-500). Our hypothesis is, these 500 users were hitting error-cases but instead of being shown feedback of what went wrong, they assumed account creation was broken and then [the potential patron] bounced.

This also tells us, a shocking 1/5 of users (i.e. 500 a day) are hitting issues when creating an account (either email or username taken, or captcha broken).

### PR DESCRIPTION
**Pull request by tabshaikh**
Issue: #3256 

Closes #1433

Todo:

- [x]  change api to check email id and screen name instead of hardwiring it to a true/false
- [x]  shift js to its respective directory and ask @jdlrobson for correct placement for js
- [x] change the function to use blur

### CODE REVIEW

**Comment by cdrini** [comment]
Let's not do this on keyup; it creates a bit of a disruptive experience to get errors thrown at you are you're typing. Let's do this on blur instead. I checked Google, Apple, Adobe, Amazon, Spotify; none of them give feedback as you type.

File: openlibrary/templates/account/create.html [hunk 1]
```diff
@@ -12,12 +12,67 @@ <h1>$_("Sign Up")</h1>
 </div>
 
 <div id="contentBody">
-
 <script type="text/javascript">
 window.q.push( function(){
```
Comment (jdlrobson) [inline openlibrary/templates/account/create.html:16]: Move this JS to a new file in openlibrary/plugins/openlibrary/js/ and then load it conditionally when #username is present in the DOM. See the existing line `if (document.getElementsByClassName('editions-table--progressively-enhanced').length) {` for an example.

It will likely help to move this code first before pushing ahead with the feature changes.

File: openlibrary/templates/account/create.html [hunk 2]
```diff
@@ -12,12 +12,67 @@ <h1>$_("Sign Up")</h1>
...
     \$('#username').keyup(function(){
```
Comment (jdlrobson) [inline openlibrary/templates/account/create.html:17]: This should be `on('keyup'`. This will break with the latest upgrade. Also best to pull this out into a separate commit following the above one.

Longer term you might want to consider using the input event which works with autofill and copy/paste.

**Comment by tabshaikh** [comment]
> Let's not do this on keyup; it creates a bit of a disruptive experience to get errors thrown at you are you're typing. Let's do this on blur instead. I checked Google, Apple, Adobe, Amazon, Spotify; none of them give feedback as you type. Also validate before the form is submitted.

@cdrini was planning to add throttling/debouncing like we have in search, I have a few issues with blur which we can discuss on the call :)

**Comment by cdrini** [comment]
:+1: Let's talk about it tomorrow!

File: openlibrary/templates/account/create.html (hunk 1 above)
Comment (tabshaikh) [reply openlibrary/templates/account/create.html]: Done :)

File: openlibrary/templates/account/create.html (hunk 2 above)
Comment (tabshaikh) [reply openlibrary/templates/account/create.html]: As discussed in the meeting with @mek and @cdrini we are planning to use `on('blur'` and I think blur should work with autofill and copy/paste as the user will inevitably click out the input box

File: openlibrary/plugins/upstream/account.py [hunk 3]
```diff
@@ -402,6 +403,45 @@ def GET(self):
         # Show failed message without thinking.
         return render['account/verify/failed']()
 
+class real_time_email_verification(delegate.page):
+    path = "/account/email/rtverify"
+    
+    def GET(self):
+        i = web.input(email='')
```
Comment (cclauss) [inline openlibrary/plugins/upstream/account.py:410]: Can we please come up with a more self-documenting variable name than `i`?

File: openlibrary/plugins/upstream/account.py [hunk 4]
```diff
@@ -402,6 +403,45 @@ def GET(self):
...
+        result = {
+                'output': False,
+                'message': 'Must be a valid email address'
+            }
+        if(bool(re.match('.*@.*', i.email))):
```
Comment (cclauss) [inline openlibrary/plugins/upstream/account.py:415]:
```suggestion
        if re.match('.*@.*', i.email):
```

File: openlibrary/plugins/upstream/account.py [hunk 5]
```diff
@@ -402,6 +403,45 @@ def GET(self):
...
+            ol_account = OpenLibraryAccount.get(email=i.email)
+            if(not(ol_account)):
+                result.output = True
+                result.message = "Email is available"
+            else:
+                result.message = "Email already registered"
```
Comment (cclauss) [inline openlibrary/plugins/upstream/account.py:421]:
```suggestion
            if ol_account:
                result.message = "Email already registered"
            else:
                result.message = "Email is available"
                result.output = True
```

File: openlibrary/plugins/upstream/account.py [hunk 6]
```diff
@@ -402,6 +403,45 @@ def GET(self):
...
+        return delegate.RawText(simplejson.dumps(result),
+                                content_type="application/json")
+
+class real_time_screenName_verification(delegate.page):
+    path = "/account/sceenname/rtverify"
+    
+    def GET(self):
+        i = web.input(name='')
+        username = i.name
```
Comment (cclauss) [inline openlibrary/plugins/upstream/account.py:430]:
```suggestion
        username = web.input(name='').name
```

File: openlibrary/plugins/upstream/account.py [hunk 7]
```diff
@@ -402,6 +403,45 @@ def GET(self):
...
+        result = {
+                'output': False,
+                'message': 'Letters and numbers only please, and at least 3 characters.'
+            }
+        if(username.isalnum() and len(username)>=3):
```
Comment (cclauss) [inline openlibrary/plugins/upstream/account.py:435]:
```suggestion
        if username.isalnum() and len(username)>=3:
```

File: openlibrary/plugins/upstream/account.py [hunk 8]
```diff
@@ -402,6 +403,45 @@ def GET(self):
...
+            ol_account = OpenLibraryAccount.get(username=username)
+            if(ol_account):
```
Comment (cclauss) [inline openlibrary/plugins/upstream/account.py:437]:
```suggestion
            if ol_account:
```

File: openlibrary/plugins/upstream/account.py (hunk 7 above)
Comment (tabshaikh) [reply openlibrary/plugins/upstream/account.py:434]: Done

Comment (cclauss) [reply openlibrary/plugins/upstream/account.py:434]: Not done.

Comment (tabshaikh) [reply openlibrary/plugins/upstream/account.py]: Ohh thought I included this in the last pr. Now done, thanks :)

File: openlibrary/plugins/openlibrary/js/realtime_account_validation.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    $('#username').on('blur', validateUsername);
+    $('#emailAddr').on('blur', validateEmail);
+    $('#password, #password2').on('blur', validatePasswords);
+
+    $('#signup').on('click', function(e) {
+        e.preventDefault();
+        if (! (window.grecaptcha && window.grecaptcha.getResponse().length)) {
+            alert('Please complete all fields and click the reCAPTCHA checkbox before proceeding.');
```
Comment (tfmorris) [inline openlibrary/plugins/openlibrary/js/realtime_account_validation.js:89]: Missing I18N

File: openlibrary/plugins/upstream/account.py [hunk 10]
```diff
@@ -402,6 +403,43 @@ def GET(self):
         # Show failed message without thinking.
         return render['account/verify/failed']()
 
+class account_validation(delegate.page):
+    path = '/account/validate'
+
+    @staticmethod
+    def validate_username(username):
+        if not 3 <= len(username) <= 20:
+            return 'Username must be between 3-20 characters'
```
Comment (tfmorris) [inline openlibrary/plugins/upstream/account.py:412]: Missing i18n

File: openlibrary/plugins/upstream/forms.py [hunk 11]
```diff
@@ -28,7 +28,7 @@ def find_ia_account(email=None):
 email_domain_not_blocked = Validator(_("Your email provider is not recognized."), lambda email: not spamcheck.is_spam_email(email))
 username_validator = Validator(_("Username already used"), lambda username: not find_account(lusername=username.lower()))
 
-vlogin = RegexpValidator(r"^[A-Za-z0-9-_]{3,20}$", _('Must be between 3 and 20 letters and numbers'))
+vlogin = RegexpValidator(r"^[A-Za-z0-9-_]{3,20}$", _('Letters and numbers only please, and at least 3 characters.'))
```
Comment (tfmorris) [inline openlibrary/plugins/upstream/forms.py:31]: This is less accurate than before (missing >20 case) and changing it causes additional downstream translation work.

File: openlibrary/templates/account/create.html [hunk 12]
```diff
@@ -50,10 +39,10 @@ <h1>$_("Sign Up")</h1>
         $if form.note:
             <div class="note">$form.note</div>
 
-        $def screenname_url(): https://openlibrary.org/people/<span id="userUrl">$(form.username.value or _('screenname'))</span>
+        $def screenname_url(): Your URL: https://openlibrary.org/people/<span id="userUrl">$(form.username.value or _('screenname'))</span>
```
Comment (tfmorris) [inline openlibrary/templates/account/create.html:42]: i18n

File: openlibrary/plugins/upstream/account.py [hunk 13]
```diff
@@ -402,6 +403,43 @@ def GET(self):
...
+        if not re.match('^[A-Za-z0-9-_]{3,20}$', username):
+            return 'Username may only contain numbers and letters'
```
Comment (tfmorris) [inline openlibrary/plugins/upstream/account.py:414]: i18n

File: openlibrary/plugins/upstream/account.py [hunk 14]
```diff
@@ -402,6 +403,43 @@ def GET(self):
...
+        ol_account = OpenLibraryAccount.get(username=username)
+        if ol_account:
+            return "Username unavailable"
```
Comment (tfmorris) [inline openlibrary/plugins/upstream/account.py:417]: i18n

File: openlibrary/plugins/upstream/account.py [hunk 15]
```diff
@@ -402,6 +403,43 @@ def GET(self):
...
+
+    @staticmethod
+    def validate_email(email):
+        if not (email and re.match('.*@.*\..*', email)):
+            return 'Must be a valid email address'
```
Comment (tfmorris) [inline openlibrary/plugins/upstream/account.py:422]: i18n

File: openlibrary/plugins/upstream/account.py [hunk 16]
```diff
@@ -402,6 +403,43 @@ def GET(self):
...
+
+        ol_account = OpenLibraryAccount.get(email=email)
+        if ol_account:
+            return 'Email already registered'
```
Comment (tfmorris) [inline openlibrary/plugins/upstream/account.py:426]: i18n

File: openlibrary/plugins/openlibrary/js/realtime_account_validation.js [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+            $('label[for="emailAddr"]').removeClass();
+            $(document.getElementById('emailAddr')).removeClass().addClass('required');
+            $('#emailAddrMessage').removeClass().text('');
+        }
+    }
+
+    function validatePasswords() {
+        var value = document.getElementById('password').value;
+        var value2 = document.getElementById('password2').value;
+        if (!value2=='' && !value=='') {
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/js/realtime_account_validation.js:63]:
```suggestion
        if (value && value2 ) {
```

File: openlibrary/plugins/openlibrary/js/realtime_account_validation.js (hunk 9 above)
Comment (mekarpeles) [reply openlibrary/plugins/openlibrary/js/realtime_account_validation.js:89]: Should this be...

```suggestion
            alert(ugettext('Please complete all fields and click the reCAPTCHA checkbox before proceeding.'));
```

?

File: openlibrary/plugins/upstream/account.py (hunk 10 above)
Comment (mekarpeles) [inline openlibrary/plugins/upstream/account.py:412]:
```suggestion
            return _('Username must be between 3-20 characters')
```

File: openlibrary/plugins/upstream/account.py (hunk 13 above)
Comment (mekarpeles) [inline openlibrary/plugins/upstream/account.py:414]:
```suggestion
            return _('Username may only contain numbers and letters')
```

File: openlibrary/plugins/upstream/account.py (hunk 15 above)
Comment (mekarpeles) [inline openlibrary/plugins/upstream/account.py:422]:
```suggestion
            return _('Must be a valid email address')
```

File: openlibrary/plugins/upstream/forms.py (hunk 11 above)
Comment (mekarpeles) [inline openlibrary/plugins/upstream/forms.py:31]:
```suggestion
vlogin = RegexpValidator(r"^[A-Za-z0-9-_]{3,20}$", _('Must be between 3 and 20 letters and numbers'))
```

File: openlibrary/plugins/upstream/account.py [hunk 18]
```diff
@@ -402,6 +403,43 @@ def GET(self):
...
 
+class account_validation(delegate.page):
+    path = '/account/validate'
+
+    @staticmethod
+    def validate_username(username):
+        if not 3 <= len(username) <= 20:
+            return _('Username must be between 3-20 characters')
+        if not re.match('^[A-Za-z0-9-_]{3,20}$', username):
+            return 'Username may only contain numbers and letters'
```
Comment (mekarpeles) [inline openlibrary/plugins/upstream/account.py]:
```suggestion
            return _('Username may only contain numbers and letters')
```

File: openlibrary/plugins/upstream/account.py [hunk 19]
```diff
@@ -402,6 +403,43 @@ def GET(self):
...
+
+    @staticmethod
+    def validate_username(username):
+        if not 3 <= len(username) <= 20:
+            return _('Username must be between 3-20 characters')
+        if not re.match('^[A-Za-z0-9-_]{3,20}$', username):
+            return _('Username may only contain numbers and letters')
+        ol_account = OpenLibraryAccount.get(username=username)
+        if ol_account:
+            return "Username unavailable"
```
Comment (mekarpeles) [inline openlibrary/plugins/upstream/account.py:417]:
```suggestion
            return _("Username unavailable")
```

File: openlibrary/plugins/upstream/account.py [hunk 20]
```diff
@@ -402,6 +403,43 @@ def GET(self):
...
+
+    @staticmethod
+    def validate_email(email):
+        if not (email and re.match('.*@.*\..*', email)):
+            return _('Must be a valid email address')
+
+        ol_account = OpenLibraryAccount.get(email=email)
+        if ol_account:
+            return 'Email already registered'
```
Comment (mekarpeles) [inline openlibrary/plugins/upstream/account.py:426]:
```suggestion
            return _('Email already registered')
```

File: openlibrary/templates/account/create.html (hunk 12 above)
Comment (mekarpeles) [inline openlibrary/templates/account/create.html:42]:
```suggestion
        $def screenname_url(): _('Your URL'): https://openlibrary.org/people/<span id="userUrl">$(form.username.value or _('screenname'))</span>
```

File: openlibrary/plugins/openlibrary/js/realtime_account_validation.js [hunk 21]
```diff
@@ -0,0 +1,97 @@
+export function initRealTimeValidation() {
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/js/realtime_account_validation.js:1]: We need to import ugettext in order to use it here

```suggestion
import { ugettext } from './i18n';

export function initRealTimeValidation() {
```

File: openlibrary/templates/account/create.html [hunk 22]
```diff
@@ -50,10 +39,10 @@ <h1>$_("Sign Up")</h1>
         $if form.note:
             <div class="note">$form.note</div>
 
-        $def screenname_url(): https://openlibrary.org/people/<span id="userUrl">$(form.username.value or _('screenname'))</span>
+        $def screenname_url(): _('Your URL'): https://openlibrary.org/people/<span id="userUrl">$(form.username.value or _('screenname'))</span>
```
Comment (mekarpeles) [inline openlibrary/templates/account/create.html:42]:
```suggestion
        $def screenname_url(): $_('Your URL'): https://openlibrary.org/people/<span id="userUrl">$(form.username.value or _('screenname'))</span>
```

File: openlibrary/plugins/openlibrary/js/realtime_account_validation.js [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+            $('label[for="emailAddr"]').removeClass();
+            $(document.getElementById('emailAddr')).removeClass().addClass('required');
+            $('#emailAddrMessage').removeClass().text('');
+        }
+    }
+
+    function validatePasswords() {
+        var value = document.getElementById('password').value;
+        var value2 = document.getElementById('password2').value;
+        if (value && value2 ) {
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/js/realtime_account_validation.js:65]:
```suggestion
        if (value && value2) {
```

File: openlibrary/plugins/openlibrary/js/realtime_account_validation.js [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+            $(document.getElementById('username')).removeClass().addClass('required');
+            $('#usernameMessage').removeClass().html('<br>');
+        }
+    }
+
+    function validateEmail() {
+        var value_email = $(this).val();
+        if (!value_email=='') {
+            $.ajax({
+                url: `/account/validate?email=${value_email}`,
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/js/realtime_account_validation.js:40]: We may have to urlencode `+` for the email. Otherwise, the API call will treat it as a space. This would means we would also have to urlencode the email on the other side.

`encodeURIComponent` from js (before we do the ajax request)
`unquote_plus` on the backend (within API)

Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/js/realtime_account_validation.js:40]:
```suggestion
                url: `/account/validate?email=${encodeURIComponent(value_email)}`,
```

File: openlibrary/plugins/upstream/account.py [hunk 25]
```diff
@@ -402,6 +403,43 @@ def GET(self):
...
         return render['account/verify/failed']()
 
+class account_validation(delegate.page):
+    path = '/account/validate'
+
+    @staticmethod
+    def validate_username(username):
+        if not 3 <= len(username) <= 20:
+            return _('Username must be between 3-20 characters')
+        if not re.match('^[A-Za-z0-9-_]{3,20}$', username):
```
Comment (mekarpeles) [inline openlibrary/plugins/upstream/account.py:413]: We should try to DRY this regex since we're now using it in 2 different places.

File: openlibrary/plugins/upstream/account.py [hunk 26]
```diff
@@ -6,6 +6,7 @@
 import datetime
 import time
 import simplejson
+import re
```
Comment (mekarpeles) [inline openlibrary/plugins/upstream/account.py:9]:
```suggestion
import re
from six.moves.urllib.parse import unquote_plus
```

File: openlibrary/plugins/upstream/account.py [hunk 27]
```diff
@@ -402,6 +403,43 @@ def GET(self):
...
+
+
+    def GET(self):
+        i = web.input()
+        errors = {
+            'email': None,
+            'username': None
+        }
+        if i.get('email') is not None:
+            errors['email'] = self.validate_email(i.email)
```
Comment (mekarpeles) [inline openlibrary/plugins/upstream/account.py:436]:
```suggestion
            errors['email'] = self.validate_email(unquote_plus(i.email))
```

File: openlibrary/plugins/upstream/account.py [hunk 28]
```diff
@@ -402,6 +403,43 @@ def GET(self):
...
+
+    def GET(self):
+        i = web.input()
+        errors = {
+            'email': None,
+            'username': None
+        }
+        if i.get('email') is not None:
+            errors['email'] = self.validate_email(unquote_plus(i.email))
+        if i.get('username') is not None:
```
Comment (cclauss) [inline openlibrary/plugins/upstream/account.py:438]:
```suggestion
        if i.get('email'):
            errors['email'] = self.validate_email(unquote_plus(i.email))
        if i.get('username'):
```
Deal properly with all _falsey_ such as None, "", {}, [], (,), False...
In the near future we may swap out web.py for more a modern frame that may return slightly different output.

File: openlibrary/plugins/upstream/account.py [hunk 29]
```diff
@@ -6,6 +6,8 @@
 import datetime
 import time
 import simplejson
+import re
+from six.moves.urllib.parse import unquote_plus
```
Comment (mekarpeles) [inline openlibrary/plugins/upstream/account.py:10]:
```suggestion
```

File: openlibrary/plugins/upstream/account.py [hunk 30]
```diff
@@ -402,6 +404,43 @@ def GET(self):
...
+
+
+    def GET(self):
+        i = web.input()
+        errors = {
+            'email': None,
+            'username': None
+        }
+        if i.get('email') is not None:
+            errors['email'] = self.validate_email(unquote_plus(i.email))
```
Comment (mekarpeles) [inline openlibrary/plugins/upstream/account.py:437]:
```suggestion
            errors['email'] = self.validate_email(i.email)
```

File: openlibrary/plugins/upstream/account.py (hunk 28 above)
Comment (mekarpeles) [reply openlibrary/plugins/upstream/account.py]: In this specific case, if e.g. an empty string `''` is provided, we want to enter this block because the API is being passed an empty email. 

e.g. https://dev.openlibrary.org/account/validate?email= should return invalid email

**Comment by cclauss** [comment]
Nice!
